flink-runtime/src/main/java/org/apache/flink/runtime/persistence/IntegerResourceVersion.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.persistence;
+
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
+
+/**
+ * {@link ResourceVersion} implementation with {@link Integer} value. The resource version in
+ * ZooKeeper is {@link Integer}.
+ */
+public class IntegerResourceVersion implements ResourceVersion, Comparable<IntegerResourceVersion> {
+
+	private static final IntegerResourceVersion NOT_EXIST = new IntegerResourceVersion(-1);
+
+	private final int value;
+
+	private IntegerResourceVersion(int value) {
+		this.value = value;
+	}
+
+	@Override
+	public int compareTo(@Nonnull IntegerResourceVersion other) {
+		return Integer.compare(value, other.getValue());
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (obj == this) {
+			return true;
+		} else if (obj != null && obj.getClass() == IntegerResourceVersion.class) {
+			final IntegerResourceVersion that = (IntegerResourceVersion) obj;
+			return this.value == that.getValue();
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public int hashCode() {
+		return Integer.hashCode(value);
+	}
+
+	@Override
+	public boolean isNotExist() {
+		return this == NOT_EXIST;
+	}
+
+	@Override
+	public String toString() {
+		return "IntegerResourceVersion{" + "value='" + value + '\'' + '}';
+	}
+
+	public int getValue() {
+		return this.value;
+	}
+
+	public static IntegerResourceVersion notExist() {
+		return NOT_EXIST;
+	}
+
+	public static IntegerResourceVersion valueOf(int value) {
+		Preconditions.checkArgument(value != NOT_EXIST.getValue());
```

Comment:
I think we should check `value >= 0`. Currently it looks like we can accept any negative value except for `-1`.
Might also mention this contract in the JavaDoc of this class.